subsys/pm/pm.c

```diff
@@ -255,23 +305,62 @@ bool pm_system_suspend(int32_t ticks)
 	return true;
 }
 
-void pm_notifier_register(struct pm_notifier *notifier)
+int pm_notifier_register(struct pm_notifier *notifier, enum pm_state state,
+			  uint8_t substate_id)
 {
-	k_spinlock_key_t pm_notifier_key = k_spin_lock(&pm_notifier_lock);
+	k_spinlock_key_t pm_notifier_key;
+	sys_slist_t *notifier_list;
+	uint8_t id = CURRENT_CPU;
+
```

Comment:
I've added a comment about registering per cpu, but here is another possible problem. In SMP you don't know which cpu you are running (Unless you have been pinned). So, the caller would need to check if it is the right CPU, sets a spin lock and call it ...  It is even worse if you are not running from the CPU you wants to get notification

Reply:
TBF I don't have a very deep understanding of how SMP works in Zephyr, since it seems a bit different from Linux with the MCU configurations of e.g. M4 + M7 cores etc.

From a quick glance I can see that Zephyr on x86 with SMP will in fact run across multiple CPUs. Do you know if this is the same on systems consisting of an M4 and M0+ core, for instance? My design tried to cover these cases, as the different cores may have diverging `cpu-power-states`, Zephyr may be pinned to the main M4 core, etc. (These are probably very flawed assumptions).

Reply:
if the core does not have a state, it will not be triggered during suspension.

Reply:
Understood, thanks for the clarification. Functionality is now CPU-independent.